### PR TITLE
[Review] Feature: Balanced-OPE estimators

### PR DESCRIPTION
**Pull request by fullflu**
## Overview

- Add Balanced OPE (B-OPE) estimators

## Tasks

- [x] initial implementation of a self-normalized balanced ips estimator (to check the behavior of my naive implementation and consider the direction of final implementation)
- [x] add cross fitting
- [x] fix action sampling procedure
- [x] add pscore estimator
- [x] add `estimated_pscore` to ope estimators
- [x] add `importance_sampling_ratio` to ope estimators
- [x] fix meta.py
- [x] add calibration
- [x] evaluate `evaluation_policy_sample_method` using various types of classifiers and hyperparameters
- [x] create evaluation notebook
- [x] request a 1st review and reflect the response
- [ ] add balanced dr estimators and report the behavior of several B-OPE estimators
- [x] fix existing tests
- [x] add new tests
- [x] request a 2nd review and reflect the response

### Self-normalized Balanced IPS

It is not trivial how to handle action features of a stochastic evaluation policy. I implement two types of method by `evaluation_policy_sample_method`.
1. `raw`: `action_dist_at_position` are directly encoded as action features.
2. `sample`: actions are sampled.
~~3. `weighted_loss`: for each round, `n_actions` rows are duplicated. For each duplicated row, action features are represented by one-hot encoding of each action. Classification models are trained with `sample_weight`, where `sample_weight` is the probability that the corresponding action is sampled (`action_dist_at_position[:, action_idx]`).~~

## Reference

https://arxiv.org/abs/1906.03694

### CODE REVIEW

**Comment by fullflu** [comment]
## Observation

Initial implementation of a self-normalized balanced ipe estimator tells us the following fact (I will commit a notebook later):
1. When the evaluation policy is deterministic, three types of `evaluation_policy_sample_method` works as good as existing estimators such as IPS and DR (`raw` and `sample` result in the same output).
2. When the evaluation policy is stochastic, `raw` does not work well (`inf`) and `sample` leads to poor performance. However, `weighted_loss` has performance comparable to existing estimators.

```
OPE of deterministic evaluation policy (IPWLearner)
- Tree: 0.6205135606005601
- raw: 0.633716233203124
- sample: 0.633716233203124
- weighted_loss: 0.6312631775591353
- IPS: 0.703706
- DR: 0.604306

OPE of stochastic evaluation policy (IPW Learner)
- True: 0.5025620586676971
- raw: nan
- sample: 0.5894081429941057
- weighted_loss: 0.506395941886161
- IPS: 0.509074
- DR: 0.506203
```

## Discussion
-  The reason why `raw` does not work well would be because the domain of action features of factual row (`\in {0, 1}`) are different with that of stochastic evaluation policy (`\in [0, 1]`). In fact, when we use `RandomForestClassifier`, the result of factual rows are all zero, and that of counterfactual rows are all one.
- The reason why `sample` leads to poor performance would be because a sampling procedure discard information of evaluation policy distribution.


## Direction

- When evaluation policy is stochastic, `weighted_loss` should be used.
- When evaluation policy is deterministic, `raw` can be used.
- `sample` seems to be unnecessary
- We should implement `ClassificationModel` class in `classification_model.py` and simplify estimators.py

**Comment by fullflu** [comment]
OPE script (stochastic policy)

```
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

# import open bandit pipeline (obp)
from obp.dataset import (
    SyntheticBanditDataset,
    logistic_reward_function,
    linear_behavior_policy,
)

from obp.policy import IPWLearner
from obp.ope import (
    OffPolicyEvaluation,
    RegressionModel,
    InverseProbabilityWeighting as IPS,
    SelfNormalizedInverseProbabilityWeighting as SNIPS,
    DirectMethod as DM,
    DoublyRobust as DR,
    DoublyRobustWithShrinkage as DRos,
    BalancedInverseProbabilityWeighting as BIPW,
)


# define a dataset class
n_actions = 10
dim_context = 8
len_list = 1
random_state = 12345
dataset = SyntheticBanditDataset(
    n_actions=n_actions,
    dim_context=dim_context,
    beta=0.2,
    reward_function=logistic_reward_function,
    behavior_policy_function=linear_behavior_policy,
    random_state=random_state,
)

# training data is used to train an evaluation policy
train_bandit_data = dataset.obtain_batch_bandit_feedback(n_rounds=5000)

# test bandit data is used to approximate the ground-truth policy value
test_bandit_data = dataset.obtain_batch_bandit_feedback(n_rounds=100000)

# evaluation policy training
ipw_learner = IPWLearner(
    n_actions=dataset.n_actions,
    base_classifier=RandomForestClassifier(random_state=random_state),
)
ipw_learner.fit(
    context=train_bandit_data["context"],
    action=train_bandit_data["action"],
    reward=train_bandit_data["reward"],
    pscore=train_bandit_data["pscore"],
)
action_dist_ipw_train = ipw_learner.predict_proba(
    context=train_bandit_data["context"],
)
action_dist_ipw_test = ipw_learner.predict_proba(
    context=test_bandit_data["context"],
)
policy_value_of_ipw = dataset.calc_ground_truth_policy_value(
    expected_reward=test_bandit_data["expected_reward"],
    action_dist=action_dist_ipw_test,
)

num_data = 1000

validation_bandit_data = dataset.obtain_batch_bandit_feedback(n_rounds=num_data)

# make decisions on vlidation data
action_dist_ipw_val = ipw_learner.predict_proba(
    context=validation_bandit_data["context"],
)

# OPE using validation data
regression_model = RegressionModel(
    n_actions=dataset.n_actions,
    base_model=LogisticRegression(C=100, max_iter=10000, random_state=random_state),
)
estimated_rewards = regression_model.fit_predict(
    context=validation_bandit_data["context"],  # context; x
    action=validation_bandit_data["action"],  # action; a
    reward=validation_bandit_data["reward"],  # reward; r
    n_folds=2,  # 2-fold cross fitting
    random_state=12345,
)

bipw = BIPW(
    estimator_name="BIPW",
    len_list=len_list,
    n_actions=n_actions,
    fit_random_state=random_state,
    base_model=RandomForestClassifier(random_state=random_state),
)

bipw.fit(
    action=train_bandit_data["action"],
    context=train_bandit_data["context"],
    action_dist=action_dist_ipw_train,
    evaluation_policy_sample_method="raw",
    position=train_bandit_data["position"],
)

pv = bipw.estimate_policy_value(
    reward=validation_bandit_data["reward"],
    action=validation_bandit_data["action"],
    position=validation_bandit_data["position"],
    action_dist=action_dist_ipw_val,
    context=validation_bandit_data["context"],
)

bipw.fit(
    action=train_bandit_data["action"],
    context=train_bandit_data["context"],
    action_dist=action_dist_ipw_train,
    evaluation_policy_sample_method="weighted_loss",
    position=train_bandit_data["position"],
)

pv_wl = bipw.estimate_policy_value(
    reward=validation_bandit_data["reward"],
    action=validation_bandit_data["action"],
    position=validation_bandit_data["position"],
    action_dist=action_dist_ipw_val,
    context=validation_bandit_data["context"],
)

bipw.fit(
    action=train_bandit_data["action"],
    context=train_bandit_data["context"],
    action_dist=action_dist_ipw_train,
    evaluation_policy_sample_method="sample",
    position=train_bandit_data["position"],
)

pv_sample = bipw.estimate_policy_value(
    reward=validation_bandit_data["reward"],
    action=validation_bandit_data["action"],
    position=validation_bandit_data["position"],
    action_dist=action_dist_ipw_val,
    context=validation_bandit_data["context"],
)

ope = OffPolicyEvaluation(
    bandit_feedback=validation_bandit_data,
    ope_estimators=[
        IPS(estimator_name="IPS"),
        DM(estimator_name="DM"),
        IPS(lambda_=5, estimator_name="CIPS"),
        SNIPS(estimator_name="SNIPS"),
        DR(estimator_name="DR"),
        DRos(lambda_=500, estimator_name="DRos"),
    ],
)


squared_errors = ope.evaluate_performance_of_estimators(
    ground_truth_policy_value=policy_value_of_ipw,  # V(\pi_e)
    action_dist=action_dist_ipw_val,  # \pi_e(a|x)
    estimated_rewards_by_reg_model=estimated_rewards,  # \hat{q}(x,a)
    metric="se",  # squared error
)

ope_res = ope.summarize_off_policy_estimates(
    action_dist=action_dist_ipw_val,  # \pi_e(a|x)
    estimated_rewards_by_reg_model=estimated_rewards,  # \hat{q}(x,a)
)

print(squared_errors)
print(ope_res)
print(policy_value_of_ipw)
print(pv, pv_sample, pv_wl)
```

**Comment by fullflu** [comment]
I remove `weighted_loss` option because classifier with `weighted_loss` is not well-trained (in terms of ROC-AUC and calibration plot of `eval_result`.)

File: obp/ope/estimators.py [hunk 1]
```diff
@@ -1731,3 +1806,239 @@ def _estimate_mse_score(
...
+        estimated_rewards: array-like or Tensor, shape (n_rounds,)
+            Rewards of each round estimated by Balanced IPW.
+
+        """
+        if position is None:
+            position = np.zeros(action_dist.shape[0], dtype=int)
+        iw = (
+            action_dist[np.arange(action.shape[0]), action, position]
+            * importance_sampling_ratio
+        )
```
Comment (usaito) [inline obp/ope/estimators.py:1905]:
```suggestion
        iw = estimated_importance_weights
```

Comment (usaito) [reply obp/ope/estimators.py:1905]: [important] in my understanding after rechecking the reference paper, \hat{\rho} in the paper estimated \pi_e / \pi_b, so we don't have to take the marginalization over the eval policy here.

File: obp/ope/estimators.py [hunk 2]
```diff
@@ -1731,3 +1806,239 @@ def _estimate_mse_score(
...
+            Ratio of probability that the action is sampled by evaluation policy divided by probability that the action is sampled by behavior policy,
+            i.e., :math:`\\hat{\\rho}(x_t, a_t)`.
+
+        action_dist: array-like, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities of evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_t|x_t)`.
+
+        position: array-like, shape (n_rounds,), default=None
+            Position of recommendation interface where action was presented in each round of the given logged bandit data.
+            When None is given, the effect of position on the reward will be ignored.
+            (If only one action is chosen and there is no posion, then you can just ignore this argument.)
```
Comment (usaito) [inline obp/ope/estimators.py:1941]:
```suggestion
            (If only a single action is chosen at each round, you can just ignore this argument.)
```

**Comment by usaito** [comment]
@fullflu Thanks again for the great work! 

The most important point I want to mention here is the definition of B-IPW.
In my understanding, we can simply use the importance weights estimated by the classification, and do not have to (re-)take the expectation over the evaluation policy in the estimator (as I made a suggestion above).

File: obp/ope/classification_model.py [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+        ----------
+        context: array-like, shape (n_rounds, dim_context)
+            Context vectors observed in each round of the logged bandit feedback, i.e., :math:`x_t`.
+
+        action: array-like, shape (n_rounds,)
+            Action sampled by behavior policy in each round of the logged bandit feedback, i.e., :math:`a_t`.
+
+        action_dist: array-like, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities of evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_t|x_t)`.
+            When either of 'iw' or 'mrdr' is used as the 'fitting_method' argument, then `action_dist` must be given.
```
Comment (fullflu) [inline obp/ope/classification_model.py:117]:
```suggestion
```

File: obp/ope/classification_model.py [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
```
Comment (fullflu) [inline obp/ope/classification_model.py:248]:
```suggestion
```

File: obp/ope/classification_model.py [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+                    f"position elements must be smaller than len_list, but the maximum value is {position.max()} (>= {self.len_list})"
+                )
+
+        check_scalar(n_folds, "n_folds", int, min_val=1)
+        check_random_state(random_state)
+
+        if not (isinstance(action_dist, np.ndarray) and action_dist.ndim == 3):
+            raise ValueError(
+                "when fitting_method is either 'iw' or 'mrdr', action_dist (a 3-dimensional ndarray) must be given"
+            )
```
Comment (fullflu) [inline obp/ope/classification_model.py:298]:
```suggestion
```

File: obp/ope/meta.py [hunk 6]
```diff
@@ -171,6 +224,19 @@ def estimate_policy_values(
             When a dict is given, if the dict has the name of a estimator as a key, the corresponding value is used.
             When it is not given, model-dependent estimators such as DM and DR cannot be used.
 
+        estimated_pscore: array-like, shape (n_rounds,), default=None
+            Estimated value of action choice probabilities of behavior policy (propensity scores), i.e., :math:`\\hat{\\pi}_b(a_t|x_t)`.
+            When an array-like is given, all OPE estimators use it.
+            When a dict is given, if the dict has the name of a estimator as a key, the corresponding value is used.
+            When it is not given, model-dependent estimators such as DM and DR cannot be used.
+
+        importance_sampling_ratio: array-like or Tensor, shape (n_rounds,), default=None
```
Comment (fullflu) [inline obp/ope/meta.py:233]: to be fixed

File: obp/ope/meta.py [hunk 7]
```diff
@@ -187,12 +253,13 @@ def estimate_policy_values(
         estimator_inputs = self._create_estimator_inputs(
             action_dist=action_dist,
             estimated_rewards_by_reg_model=estimated_rewards_by_reg_model,
+            estimated_pscore=estimated_pscore,
+            importance_sampling_ratio=importance_sampling_ratio,
```
Comment (fullflu) [inline obp/ope/meta.py:257]: to be fixed

File: obp/ope/classification_model.py [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+from sklearn.utils import check_scalar
+from sklearn.calibration import CalibratedClassifierCV
+
+from ..utils import check_array, sample_action_fast
+
+
+@dataclass
+class ImportanceSampler(BaseEstimator):
+    """Machine learning model to distinguish between the behavior and evaluation policy (:math:`\\Pr[C = 1 | x, a]`),
+    where :math:`\\Pr[C=1|x,a]` is the probability that the action :math:`a` is sampled by evaluation policy given :math:`x`.
```
[the remark this answers is not in the document]
Comment (fullflu) [reply obp/ope/classification_model.py:22]: fixed!

File: obp/ope/classification_model.py [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+        Number of actions.
+
+    len_list: int, default=1
+        Length of a list of actions recommended in each impression.
+        When Open Bandit Dataset is used, 3 should be set.
+
+    action_context: array-like, shape (n_actions, dim_action_context), default=None
+        Context vector characterizing action (i.e., vector representation of each action).
+        If not given, one-hot encoding of the action variable is used as default.
+        If fitting_method is 'raw', action_context must be None.
```
Comment (usaito) [inline obp/ope/classification_model.py:39]: [IMO] We use only one-hot encoding style `action_context` in the current implementation, i.e., setting some special action contexts does not change the estimation results. Then, why not deleting this argument and use `self.action_context=self.action_context = np.eye(self.n_actions, dtype=int)`? I thought this extremely simplifies the current implementation while achieving the same procedure.

File: tests/ope/test_offline_estimation_performance.py [hunk 10]
```diff
@@ -160,6 +214,7 @@ def test_offline_estimation_performance(
     dim_context: int,
     base_model_for_evaluation_policy: str,
     base_model_for_reg_model: str,
+    base_model_for_treatment_model: str,
```
Comment (usaito) [inline tests/ope/test_offline_estimation_performance.py:217]:
```suggestion
    base_model_for_pscore_estimator: str,
```

File: tests/ope/test_offline_estimation_performance.py [hunk 11]
```diff
@@ -160,6 +214,7 @@ def test_offline_estimation_performance(
     dim_context: int,
     base_model_for_evaluation_policy: str,
```
Comment (usaito) [inline tests/ope/test_offline_estimation_performance.py:215]:
```suggestion
    base_model_for_iw_estimator: str,
```

File: obp/ope/classification_model.py (hunk 9 above)
Comment (usaito) [inline obp/ope/classification_model.py:39]:
```suggestion
        If fitting_method is 'raw', one-hot encoding will be used as action_context.
```

File: obp/ope/classification_model.py [hunk 12]
```diff
@@ -0,0 +1,660 @@
...
+
+    fitting_method: str, default='sample'
+        Method to fit the classification model.
+        Must be one of ['sample', 'raw']. Each method is defined as follows:
+            - sample: actions are sampled by applying Gumbel-softmax trick to action_dist_at_position, and action features are represented by one-hot encoding of the sampled action.
+            - raw: action_dist_at_position are directly encoded as action features.
+        If fitting_method is 'raw', action_context must be None.
```
Comment (usaito) [inline obp/ope/classification_model.py:46]:
```suggestion
        If fitting_method is 'raw', one-hot encoding will be used as action_context.
```

File: obp/ope/classification_model.py [hunk 13]
```diff
@@ -0,0 +1,660 @@
...
+                self.action_context.shape
+                != (
+                    self.n_actions,
+                    self.n_actions,
+                )
+                or not np.allclose(self.action_context, np.eye(self.n_actions))
+            ):
+                raise ValueError(
+                    "If fitting_method == 'raw', action_context must be None or identity matrix of size (n_actions, n_actions)."
+                )
```
Comment (usaito) [inline obp/ope/classification_model.py:92]:
```suggestion
```

File: obp/ope/classification_model.py [hunk 14]
```diff
@@ -0,0 +1,660 @@
...
+                        base_estimator=self.base_model, cv=self.calibration_cv
+                    ),
+                )
+                for _ in np.arange(self.len_list)
+            ]
+        else:
+            self.base_model_list = [
+                clone(self.base_model) for _ in np.arange(self.len_list)
+            ]
+        if self.action_context is None:
```
Comment (usaito) [inline obp/ope/classification_model.py:111]:
```suggestion
        if self.action_context is None or self.fitting_method == "raw":
```

Comment (usaito) [reply obp/ope/classification_model.py:111]: @fullflu I thought we can just set one-hot encoding as action context when self.fitting_method == "raw".

File: obp/ope/classification_model.py [hunk 15]
```diff
@@ -0,0 +1,660 @@
...
+        Parameters
+        -----------
+        context: array-like, shape (n_rounds,)
+            Context vectors observed in each round of the logged bandit feedback, i.e., :math:`x_t`.
+
+        action: array-like, shape (n_rounds,)
+            Action sampled by behavior policy in each round of the logged bandit feedback, i.e., :math:`a_t`.
+
+        action_context: array-like, shape shape (n_actions, dim_action_context)
+            Context vector characterizing action (i.e., vector representation of each action).
```
Comment (usaito) [inline obp/ope/classification_model.py:394]:
```suggestion
```

File: obp/ope/classification_model.py [hunk 16]
```diff
@@ -0,0 +1,660 @@
...
+
+        action_dist_at_position: array-like, shape (n_rounds, n_actions,)
+            Action choice probabilities of evaluation policy of each position (can be deterministic), i.e., :math:`\\pi_e(a_t|x_t)`.
+
+        sampled_action_at_position: array-like, shape (n_rounds, n_actions,)
+            One-hot encoding of actions sampled by evaluation policy of each position.
```
Comment (usaito) [inline obp/ope/classification_model.py:400]:
```suggestion
            Actions sampled by evaluation policy for each data at each position.
```

File: obp/ope/classification_model.py [hunk 17]
```diff
@@ -0,0 +1,660 @@
...
+
+    action_context: array-like, shape (n_actions, dim_action_context), default=None
+        Context vector characterizing action (i.e., vector representation of each action).
+        If not given, one-hot encoding of the action variable is used as default.
+        If fitting_method is 'raw', action_context must be None.
+
+    fitting_method: str, default='sample'
+        Method to fit the classification model.
+        Must be one of ['sample', 'raw']. Each method is defined as follows:
+            - sample: actions are sampled by applying Gumbel-softmax trick to action_dist_at_position, and action features are represented by one-hot encoding of the sampled action.
```
[the remark this answers is not in the document]
Comment (fullflu) [reply obp/ope/classification_model.py:44]: [fix]

action features are not restricted to one-hot encoding

**Pull request merged**